### PR TITLE
fix: use `--retry-connrefused` in curl invocations of bootstrap download

### PR DESCRIPTION
**Pull request by kayagokalp**
helps with #110178.

### CODE REVIEW

**Review by onur-ozkan — CHANGES_REQUESTED** [review]
Thanks for the PR!

`--retry-all-errors` should be an extra argument with keeping `--retry` as it is. Otherwise curl will hang the downloading processes until the timeout.

e.g:

**Comment by ozkanonur** [comment]
this lgtm, once you squash your commits, we can merge the PR

**Comment by matthiaskrgr** [comment]
@bors r- 
failed in a rollup
https://github.com/rust-lang/rust/pull/110271

**Comment by klensy** [comment]
Looks like `--retry-all-errors` supported from [curl 7.71.0](https://daniel.haxx.se/blog/2020/06/24/curl-7-71-0-blobs-and-retries/) and some runners older than that?

**Comment by ozkanonur** [comment]
> Looks like `--retry-all-errors` supported from [curl 7.71.0](https://daniel.haxx.se/blog/2020/06/24/curl-7-71-0-blobs-and-retries/) and some runners older than that?
> 
> ubuntu 20.04 image uses curl 7.68 https://packages.ubuntu.com/focal/curl

Yeah seems like `--retry-all-errors` flag shipped very recently.

@kayagokalp Can you replace the `--retry-all-errors` flag with `--retry-connrefused` and update commit message/PR description? `--retry-connrefused` added since v7.52.0, which should work on any CI/local machine.

File: src/bootstrap/bootstrap.py
Comment (ozkanonur) [inline src/bootstrap/bootstrap.py:95]:
```suggestion
                "--retry", "3", "--retry-connrefused", "-Sf", url],
```

**Comment by klensy** [comment]
centos 7 image have curl 7.29 :-(. It's easier to download our own version and install/build it.

**Comment by ozkanonur** [comment]
> centos 7 image have curl 7.29 :-(.

7.52 released in 2016, I was pretty sure that we have more recent version...

"curl version 7.29. 0 was released on February 6 2013." :(

> It's easier to download our own version and install/build it.

Doing custom build or installation just for couple new arguments, it doesn't really seem worth to me.

cc @jyn514

**Comment by jyn514** [comment]
> Doing custom build or installation just for couple new arguments, it doesn't really seem worth to me.

i agree. instead we can put the retry logic in Rust (and make sure not to retry if it's a 404).

**Comment by kayagokalp** [comment]
Going to close this for now, and re ignite once I got some time again if no one beats mo to it!

**Pull request closed without merging**
